libsolidity/ast/AST.cpp

```diff
@@ -374,6 +374,29 @@ std::vector<std::pair<ASTPointer<IdentifierPath>, std::optional<Token>>> UsingFo
 	return ranges::zip_view(m_functionsOrLibrary, m_operators) | ranges::to<vector>;
 }
 
+std::string StructDefinition::encodeType() const
+{
+	std::string str = name() + "(";
+	std::set<std::string> nested; // std::set enables duplicates elimination and ordered enumeration
+	for (size_t i = 0; i < m_members.size(); i++)
+	{
+		str += i == 0 ? "" : ",";
+		str += m_members[i]->type()->canonicalName() + " " + m_members[i]->name();
+		if (m_members[i]->type()->category() == Type::Category::Struct) {
+			Declaration const* declaration = m_members[i]->type()->typeDefinition();
+			if (StructDefinition const* structDef = dynamic_cast<StructDefinition const*>(declaration)) {
+				nested.insert(structDef->encodeType());
```

Comment:
Ok, got it, you mean we probably need to extract all the nested structs into flat list, sort it and then append. Got it, I know how to do this.

Reply:
Implemented with the following set of methods:
```
/// Fills set with the EIP-712 compatible struct encodings without subtypes concatenated.
void insertEncodedSubtypes(std::set<std::string>& subtypes) const;

/// @returns the EIP-712 compatible struct encoding but without subtypes concatenated.
std::string encodeTypeWithoutSubtypes() const;

/// @returns the EIP-712 compatible struct encoding with subtypes sorted and concatenated.
std::string encodeType() const;

/// @returns the EIP-712 compatible typehash of this struct.
util::h256 typehash() const;
```

Had an idea of method returning `std::set` and calling it recursively for subtypes, but this would require to merge sets, so I decided that filling single set would be more efficient